deploy/helm-chart/templates/deployment.yaml

```diff
@@ -104,6 +104,10 @@ spec:
           mountPath: /etc/nginx/secrets
         - name: nginx-run
           mountPath: /var/run/nginx
+        {{- if .Values.nginx.plus }}
+        - name: nginx-state
```

Comment:
I wonder if introducing a volume can affect graceful recovery https://github.com/nginxinc/nginx-gateway-fabric/blob/main/tests/graceful-recovery/graceful-recovery.md ? Will NGINX be able to successfully restart if it is not empty?

Is there a possibility of a leak here? For example,  NGINX restarts after a failure, old state file remains but the control plane generates NGINX configuration without some of those upstreams. And that repeats over time.

Also, does NGINX remove the state file after an upstream was removed from its configuration?

Reply:
1. We already have volumes for files that nginx needs access to, why would this one be different from those?
2/3. It does not appear that the state file is removed when an upstream is removed. However, the file is empty, so it takes up minimal space. There would need to be a very significant amount of empty files to make any real impact on disk usage. On a restart, the empty state file is removed.

Reply:
> We already have volumes for files that nginx needs access to, why would this one be different from those?
I didn't test that path, so that's why I'm bringing it up

Note there is a bug related to restart https://github.com/nginxinc/nginx-gateway-fabric/issues/1108

An alternative to state files - just populating upstream servers as for NGINX OSS. If we need to change them for Plus for EndpointSlice changes, we just use N+ API to change them - NGINX Plus will be ok if the list in the config doesn't match the list in its memory.

When later NGF needs to reload NGINX Plus, since it will regenerate all config files, which includes adding upstream servers to upstreams, no upstream servers will be lost after a reload.

Reply:
For debuggability, I'm not too fond of the list in the nginx.conf potentially not being an accurate representation of the actual server list. Especially in a situation where the environment is stable, as in no config changes, except for endpoint scaling; we may not get a reload for a long time.

Reply:
One alternative: For debuggability without state file , one approach could be still updating conf file, but just skipping a reload. To avoid updating all config, we can put all upstreams in a separate file

Reply:
Yeah, that could work...

Reply:
Though now we could introduce another edge case...conf is updated (no reload), and then API fails to update the servers. Now the conf and reality are out of sync.

Reply:
The nice thing with the state file is if the upstreams change, we don't have to do any template generation or file writing/removing.